component/componenterror/errors.go

```diff
@@ -48,16 +48,25 @@ func CombineErrors(errs []error) error {
 	}
 
 	errMsgs := make([]string, 0, numErrors)
-	permanent := false
+	permanent, partialScrapeErr, failedScrapeCount := false, false, 0
 	for _, err := range errs {
 		if !permanent && consumererror.IsPermanent(err) {
 			permanent = true
 		}
+		if partialError, isPartial := err.(consumererror.PartialScrapeError); isPartial {
+			partialScrapeErr = true
+			failedScrapeCount += partialError.Failed
+		}
+
 		errMsgs = append(errMsgs, err.Error())
 	}
+
 	err := fmt.Errorf("[%s]", strings.Join(errMsgs, "; "))
-	if permanent {
+	switch {
+	case permanent:
 		err = consumererror.Permanent(err)
+	case partialScrapeErr:
```

Comment:
Hmm. Feels like the specifics of partial errors don't belong in this function. Two ideas:

1. Add concept of "combinable" errors

```go
var combined Combineable
if ok, e := err.(Combineable); ok {
   combined = e.Combine(combined)
}
```

In this case if PartialError was a Combineable on first encounter of a PartialError combined is nil so e.Combine(combined) would return PartialError($num_failed), if there was a second one e.Combine(combined) would return combined.num_failed + $num_failed, and so on. Combine functions would also need to make sure it's combining like types.

This may be overcomplicated and I may be missing some details so see #2.

2. Have the scraper pre-combine/accumulate ScrapeError. Maybe have a wrapper around CombineErrors that scrapers use, something like that.

Reply:
Yea you're definitely right there. I changed this to no. 2